File: binding/src/multimedia/libqaudioinput.cpp
#include <QAudioDevice>
#include <QAudioInput>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qaudioinput.h>
#include "libqaudioinput.h"
#include "libqaudioinput.hxx"

QAudioInput* QAudioInput_new() {
	 return new QAudioInput();
}

QAudioInput* QAudioInput_new2(const QAudioDevice* deviceInfo) {
	 return new QAudioInput(*deviceInfo);
}

QAudioInput* QAudioInput_new3(QObject* parent) {
	 return new QAudioInput(parent);
}

QAudioInput* QAudioInput_new4(const QAudioDevice* deviceInfo, QObject* parent) {
	 return new QAudioInput(*deviceInfo, parent);
}

libqt_string QAudioInput_Tr(const char* s) {
	QString _ret = QAudioInput::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QAudioDevice* QAudioInput_Device(const QAudioInput* self) {
	return new QAudioDevice(self->device());
}

float QAudioInput_Volume(const QAudioInput* self) {
	return self->volume();
}

bool QAudioInput_IsMuted(const QAudioInput* self) {
	return self->isMuted();
}

void QAudioInput_SetDevice(QAudioInput* self, const QAudioDevice* device) {
	self->setDevice(*device);
}

void QAudioInput_SetVolume(QAudioInput* self, float volume) {
	self->setVolume(volume);
}

void QAudioInput_SetMuted(QAudioInput* self, bool muted) {
	self->setMuted(muted);
}

void QAudioInput_DeviceChanged(QAudioInput* self) {
	self->deviceChanged();
}

void QAudioInput_Connect_DeviceChanged(QAudioInput* self, intptr_t slot) {
    void (*slotFunc)(QAudioInput*) = reinterpret_cast<void (*)(QAudioInput*)>(slot);
    QAudioInput::connect(self, &QAudioInput::deviceChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QAudioInput_VolumeChanged(QAudioInput* self, float volume) {
	self->volumeChanged(volume);
}

void QAudioInput_Connect_VolumeChanged(QAudioInput* self, intptr_t slot) {
    void (*slotFunc)(QAudioInput*, float) = reinterpret_cast<void (*)(QAudioInput*, float)>(slot);
    QAudioInput::connect(self, &QAudioInput::volumeChanged, [self, slotFunc](float volume) {
	slotFunc(self, volume);
    });
}

void QAudioInput_MutedChanged(QAudioInput* self, bool muted) {
	self->mutedChanged(muted);
}

void QAudioInput_Connect_MutedChanged(QAudioInput* self, intptr_t slot) {
    void (*slotFunc)(QAudioInput*, bool) = reinterpret_cast<void (*)(QAudioInput*, bool)>(slot);
    QAudioInput::connect(self, &QAudioInput::mutedChanged, [self, slotFunc](bool muted) {
	slotFunc(self, muted);
    });
}

libqt_string QAudioInput_Tr2(const char* s, const char* c) {
	QString _ret = QAudioInput::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QAudioInput_Tr3(const char* s, const char* c, int n) {
	QString _ret = QAudioInput::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QAudioInput_Delete(QAudioInput* self) {
    delete self;
}

File: binding/src/multimedia/libqaudioinput.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qaudioinput.html
pub const qaudioinput = struct {

    /// New constructs a new QAudioInput object.
    pub fn New() QtC.QAudioInput {
        return qtc.QAudioInput_new();
    }


    /// New2 constructs a new QAudioInput object.
    pub fn New2(deviceInfo: ?*anyopaque) QtC.QAudioInput {
        return qtc.QAudioInput_new2(@ptrCast(deviceInfo));
    }


    /// New3 constructs a new QAudioInput object.
    pub fn New3(parent: ?*anyopaque) QtC.QAudioInput {
        return qtc.QAudioInput_new3(@ptrCast(parent));
    }


    /// New4 constructs a new QAudioInput object.
    pub fn New4(deviceInfo: ?*anyopaque, parent: ?*anyopaque) QtC.QAudioInput {
        return qtc.QAudioInput_new4(@ptrCast(deviceInfo), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAudioInput_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qaudioinput.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Device(self: ?*anyopaque, ) QtC.QAudioDevice {
        return qtc.QAudioInput_Device(@ptrCast(self));
    }

    pub fn Volume(self: ?*anyopaque, ) f32 {
        return qtc.QAudioInput_Volume(@ptrCast(self));
    }

    pub fn IsMuted(self: ?*anyopaque, ) bool {
        return qtc.QAudioInput_IsMuted(@ptrCast(self));
    }

    pub fn SetDevice(self: ?*anyopaque, device: ?*anyopaque) void {
        qtc.QAudioInput_SetDevice(@ptrCast(self), @ptrCast(device));
    }

    pub fn SetVolume(self: ?*anyopaque, volume: f32) void {
        qtc.QAudioInput_SetVolume(@ptrCast(self), volume);
    }

    pub fn SetMuted(self: ?*anyopaque, muted: bool) void {
        qtc.QAudioInput_SetMuted(@ptrCast(self), muted);
    }

    pub fn DeviceChanged(self: ?*anyopaque, ) void {
        qtc.QAudioInput_DeviceChanged(@ptrCast(self));
    }

    pub fn OnDeviceChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QAudioInput_Connect_DeviceChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn VolumeChanged(self: ?*anyopaque, volume: f32) void {
        qtc.QAudioInput_VolumeChanged(@ptrCast(self), volume);
    }

    pub fn OnVolumeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f32) callconv(.c) void) void {
        qtc.QAudioInput_Connect_VolumeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MutedChanged(self: ?*anyopaque, muted: bool) void {
        qtc.QAudioInput_MutedChanged(@ptrCast(self), muted);
    }

    pub fn OnMutedChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QAudioInput_Connect_MutedChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAudioInput_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qaudioinput.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAudioInput_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qaudioinput.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QAudioInput_Delete(@ptrCast(self));
    }
};
